pkg/cloudprovider/providers/vsphere/vclib/connection.go

```diff
@@ -123,9 +116,31 @@ func (connection *VSphereConnection) login(ctx context.Context, client *vim25.Cl
 	signer, err := tokens.Issue(ctx, req)
 	if err != nil {
 		klog.Errorf("Failed to issue SAML token. err: %+v", err)
+		return nil, err
+	}
+
+	return signer, nil
+}
+
+// login calls SessionManager.LoginByToken if certificate and private key are configured,
+// otherwise calls SessionManager.Login with user and password.
+func (connection *VSphereConnection) login(ctx context.Context, client *vim25.Client) error {
+	m := session.NewManager(client)
+	connection.credentialsLock.Lock()
+	defer connection.credentialsLock.Unlock()
+
+	signer, err := connection.Signer(ctx, client)
+	if err != nil {
 		return err
 	}
 
+	if signer == nil {
+		klog.V(3).Infof("SessionManager.Login with username '%s'", connection.Username)
```

Comment:
optionally `... username: %s`

Reply:
Do you mean remove the quotes?  It can be useful to quote values from config to see any leading/trailing whitespace.  But using `%q` is probably better for that.

Reply:
yes, i guess `%q` is the alternative to `: %s`, if whitespace is to be tracked.

Reply:
@neolit123 ok changed to `%q`